File: kimdaeyeong/src/programmers/Pg154540.java
package programmers;

import java.util.*;

/**
 * 프로그래머스 154540
 * 무인도 여행
 * lv2
 * https://school.programmers.co.kr/learn/courses/30/lessons/154540
 */
public class Pg154540 {
    private static char[][] m; // 지도 정보
    private static int r; // 지도의 행 사이즈
    private static int c; // 지도의 열 사이즈
    private static int days; // 총 머무를 수 있는 일수

    private static int[] dx = {-1, 0, 1, 0};
    private static int[] dy = {0, -1, 0, 1};

    public static void main(String[] args) {
        String[] maps = {"X591X","X1X5X","X231X", "1XXX1"};
        List<Integer> answer = new ArrayList<>();

        r = maps.length;
        c = maps[0].length();
        m = new char[r][c];

        for(int i = 0; i < r; i++)
            m[i] = maps[i].toCharArray();

        for(int i = 0; i < r; i++) {
            for(int j = 0; j < c; j++) {
                if(m[i][j] != 'X') {
                    days = Character.getNumericValue(m[i][j]);
                    m[i][j]= 'X';
                    dfs(i, j);
                    answer.add(days);
                }
            }
        }

        if(answer.isEmpty()) {
            answer.add(-1);
            System.out.println(answer);
        } else {
            Collections.sort(answer);

            System.out.println(answer);
        }
    }

    private static void dfs(int x, int y) {

        for(int i = 0; i < 4; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];

            if(nx >= r || nx < 0 || ny >= c || ny < 0)
                continue;

            if(m[nx][ny] == 'X')
                continue;

            days += Character.getNumericValue(m[nx][ny]);
            m[nx][ny] = 'X';

            dfs(nx, ny);
        }
    }
}

File: limsubin/programmers/pg154540.py
# PG 154540: 무인도 여행
# https://school.programmers.co.kr/learn/courses/30/lessons/154540

from collections import deque

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

def bfs(maps, x, y):
    queue = deque()
    queue.append([x, y])

    sum = int(maps[x][y]) # 식량의 합
    maps[x][y] = 'X' # 방문 체크

    while queue:
        x, y = queue.popleft()

        for i in range(4):
            mx = x + dx[i]
            my = y + dy[i]

            if mx < 0 or mx >= len(maps) or my < 0 or my >= len(maps[0]):
                continue

            if maps[mx][my] == 'X':
                continue

            # 연결되어 있으면 방문!
            sum += int(maps[mx][my])
            queue.append([mx, my])
            maps[mx][my] = 'X'

    return sum

def solution(maps):
    answer = []
    maps = [list(i) for i in maps] # 리스트 형태로 변환

    for i in range(len(maps)):
        for j in range(len(maps[0])):
            # 바다가 아닌 곳 방문
            if maps[i][j] != 'X':
                answer.append(bfs(maps, i, j))

    # 오름차순
    if len(answer) > 0:
        answer.sort()
    # 무인도가 없는 경우
    else:
        answer.append(-1)

    return answer